### PR TITLE
Adding RealmQuery.isNull() and RealmQuery.isNotNull().

### LINKED ISSUE #975: Querying for objects with or without relationships
It could be useful to be able to query for objects where relationships (`RealmList<...>`) are empty (or not empty).

Such a query could be:

    RealmResults attachedTires = realm.where(RealmTire.class).equalTo("car", null).findAll();

### PR DESCRIPTION
**Pull request by kneth**
Fixes #975.

@cmelchior @emanuelez

### CODE REVIEW

File: realm/src/main/java/io/realm/RealmQuery.java [hunk 1]
```diff
@@ -169,6 +169,30 @@ private boolean containsDot(String s) {
         }
     }
 
+    // isNull and isNotNull
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmQuery.java]: Having these kind of sections is not really maintainable. So I would prefer if we got rid of them.

File: realm/src/main/java/io/realm/RealmQuery.java [hunk 2]
```diff
@@ -169,6 +169,30 @@ private boolean containsDot(String s) {
...
+
+    /**
+     * Add condition "Is relationship null?"
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmQuery.java]: A bit convoluted. Perhaps "Test if a field is null. Only works for relationships and RealmLists"

File: realm/src/main/java/io/realm/RealmQuery.java [hunk 3]
```diff
@@ -169,6 +169,30 @@ private boolean containsDot(String s) {
...
+        if (containsDot(fieldName)) {
+            throw new IllegalArgumentException("Following multiple relationships is not supported.");
+        }
+
+        // checking that fieldName has the correct type is done in C++
+        this.query.isNull(columns.get(fieldName));
+        return this;
+    }
+
+    public RealmQuery<E> isNotNull(String fieldName) {
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmQuery.java]: Missing JavaDoc

File: realm-jni/src/util.hpp [hunk 4]
```diff
@@ -403,6 +419,12 @@ inline bool TblColIndexAndTypeValid(JNIEnv* env, T* pTable, jlong columnIndex, i
         && ColIndexAndTypeValid(env, pTable, columnIndex, expectColType);
 }
 
+template <class T>
+inline bool TblColIndexAndLinklike(JNIEnv* env, T* pTable, jlong columnIndex) {
+    return TableIsValid(env, pTable)
+        && TypeIsLinkLike(env, pTable, columnIndex);
```
Comment (cmelchior) [inline realm-jni/src/util.hpp]: Why newline?

File: realm/src/androidTest/java/io/realm/RealmLinkTests.java [hunk 5]
```diff
@@ -463,4 +463,29 @@ public void testSubquery() {
...
+        RealmResults<Owner> owners2 = testRealm.where(Owner.class).isNotNull("cat").findAll();
+        assertEquals(1, owners2.size());
+    }
+
+    public void testIsNullWrongType() {
+        try {
+            // Owner.name is a String
+            RealmResults<Owner> owners = testRealm.where(Owner.class).isNull("name").findAll();
+            fail();
+        } catch (IllegalArgumentException e) {
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmLinkTests.java]: e -> expected

File: realm/src/androidTest/java/io/realm/RealmLinkTests.java [hunk 6]
```diff
@@ -463,4 +463,29 @@ public void testSubquery() {
         RealmResults<Owner> subOwners = owners.where().equalTo("cat.name", "Blackie").findAll();
         assertEquals(1, subOwners.size());
     }
+
+    public void testIsNull() {
+        RealmResults<Owner> owners1 = testRealm.where(Owner.class).isNull("dogs").findAll();
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmLinkTests.java]: I think we should also test on a null link and linklist

Comment (astigsen) [inline realm/src/androidTest/java/io/realm/RealmLinkTests.java]: I don't think that a link list can ever logically be null (it can be empty
though).

On Friday, March 20, 2015, Emanuele Zattin <notifications@github.com> wrote:

> In realm/src/androidTest/java/io/realm/RealmLinkTests.java
> <https://github.com/realm/realm-java/pull/980#discussion_r26848052>:
>
> > @@ -463,4 +463,29 @@ public void testSubquery() {
> >          RealmResults<Owner> subOwners = owners.where().equalTo("cat.name", "Blackie").findAll();
> >          assertEquals(1, subOwners.size());
> >      }
> > +
> > +    public void testIsNull() {
> > +        RealmResults<Owner> owners1 = testRealm.where(Owner.class).isNull("dogs").findAll();
>
> I think we should also test on a null link and linklist
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/realm/realm-java/pull/980/files#r26848052>.
>


-- 
*Alexander Stigsen, CEO*
Realm Inc

148 Townsend St
San Fransisco, CA 94107
United States

Comment (kneth) [inline realm/src/androidTest/java/io/realm/RealmLinkTests.java]: When we get support for null, I would like to extend this predicate to a more general one.

**Comment by kneth** [comment]
I have updated a few thing. @emanuelez and @cmelchior please do a quick review before merging.

File: realm/src/main/java/io/realm/internal/TableQuery.java [hunk 7]
```diff
@@ -660,6 +660,14 @@ public Date minimumDate(long columnIndex) {
     protected native long nativeMinimumDate(long nativeQueryPtr, long columnIndex, long start, long end, long limit);
 
 
+    // isNull and isNotNull
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/internal/TableQuery.java]: Remove comment

File: realm/src/main/java/io/realm/RealmQuery.java [hunk 8]
```diff
@@ -169,6 +169,35 @@ private boolean containsDot(String s) {
         }
     }
 
+    /**
+     * Test if a field is null. Only works for relationships and RealmLists.
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/RealmQuery.java]: if -> whether

File: realm-jni/src/util.hpp (hunk 4 above)
Comment (cmelchior) [inline realm-jni/src/util.hpp]: Still wondering about this newline? :)

File: realm/src/androidTest/java/io/realm/RealmLinkTests.java [hunk 9]
```diff
@@ -463,4 +463,53 @@ public void testSubquery() {
...
+        testRealm.commitTransaction();
+
+        RealmResults<Owner> owners3 = testRealm.where(Owner.class).isNull("dogs").findAll();
+        assertEquals(1, owners3.size());
+
+        RealmResults<Owner> owners4 = testRealm.where(Owner.class).isNull("cat").findAll();
+        assertEquals(1, owners4.size());
+    }
+
+    public void testIsNotNull() {
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmLinkTests.java]: These test test both LINK and LINK_LIST. It would be much more readable and easier to see if functionality is tested by splitting it into: testRelationshipIsNotNull() and testRealmListIsNotNull() and similar for isNull()

File: realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 10]
```diff
@@ -22,6 +22,7 @@ using namespace tightdb;
 
 #if 1
 #define COL_TYPE_VALID(env,ptr,col, type)           TBL_AND_COL_INDEX_AND_TYPE_VALID(env,ptr,col, type)
+#define COL_TYPE_LINKLIKE(env,ptr,col)              TBL_AND_COL_INDEX_AND_LINKLIKE(env,ptr,col)
```
Comment (cmelchior) [inline realm-jni/src/io_realm_internal_TableQuery.cpp]: Small nitpick, but wording like his makes me wonder what it means. LINK_OR_LINKLIST would be better imo. (although a bit long)

File: realm/src/main/java/io/realm/RealmQuery.java [hunk 11]
```diff
@@ -169,6 +169,35 @@ private boolean containsDot(String s) {
...
+     *
+     * @param fieldName - the field name
+     * @return The query object
+     * @throws java.lang.IllegalArgumentException if field is not a RealmObject or RealmList
+     */
+    public RealmQuery<E> isNull(String fieldName) {
+        // Currently we only support querying top-level
+        if (containsDot(fieldName)) {
+            throw new IllegalArgumentException("Following multiple relationships is not supported.");
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmQuery.java]: "Following multiple relationships" reads a bit unclear to me, Perhaps reword it to "Checking for null in nested objects is not supported."
